testnet-contracts/contracts/Oracle.sol

```diff
@@ -0,0 +1,297 @@
+    *       Make a new OracleClaim on an existing CosmosBridgeClaim
+    */
+    function newOracleClaim(
+        uint256 _cosmosBridgeNonce,
+        address _validatorAddress,
+        bytes32 _contentHash,
+        bytes memory _signature
+    )
+        internal
+    {
+        // Create a new claim
+        OracleClaim memory oracleClaim = OracleClaim(
+            _validatorAddress,
+            _contentHash,
+            _signature,
+            true
+        );
+
+        // Add the oracle claim to this CosmosBridgeClaim's OracleClaims
+        oracleClaims[_cosmosBridgeNonce].push(oracleClaim);
+        // Add the oracle claim to this validator's OracleClaims
+        validatorOracleClaims[_validatorAddress].push(_cosmosBridgeNonce);
+
+        emit LogNewOracleClaim(
+            _cosmosBridgeNonce,
+            _validatorAddress,
+            _contentHash,
+            _signature
+        );
+    }
+
+    /*
+    * @dev: processProphecyClaim
+    *       Attempts to process a prophecy claim submission by reaching
+    *       validator consensus
+    */
+    function processProphecyClaim(
+        uint256 _cosmosBridgeClaimId,
+        address _submitter,
+        bytes32 _hash,
+        address[] memory _signers,
+        uint8[] memory _v,
+        bytes32[] memory _r,
+        bytes32[] memory _s
+    )
+        internal
+    {
+        uint256 signedPower = 0;
+
+        // Iterate over the signatory addresses
+        for (uint256 i = 0; i < _signers.length; i = i.add(1)) {
+            address signer = _signers[i];
+
+            // Only consider this signer if it's an unused address
+            if(!usedAddresses[signer]) {
+                // Mark this signer's address as used
+                usedAddresses[signer] = true;
+
+                // Validate the signature
+                bool valid = isValidSignature(
+                    signer,
+                    _hash,
+                    _v[i],
+                    _r[i],
+                    _s[i]
+                );
+
+                // Only add the power of active validators
+                if(valid && activeValidators[signer]) {
+                    signedPower = signedPower.add(powers[signer]);
+                }
+            }
+        }
+
+        // Reset usedAddresses mapping
+        for (uint256 i = 0; i < _signers.length; i = i.add(1)) {
+            if(usedAddresses[_signers[i]]) {
+                usedAddresses[_signers[i]] = false;
+            }
+        }
+
+        require(
+            signedPower.mul(3) > totalPower.mul(2),
+            "The cumulative power of signatory validators does not meet the threshold"
+        );
+
+        emit LogProphecyProcessed(
+            _cosmosBridgeClaimId,
+            signedPower,
+            totalPower,
+            _submitter
+        );
+    }
+
+    /*
+    * @dev: updateValidatorsPower
+    *       Allows the provider to update both the validators and powers
+    */
+   function updateValidatorsPower(
```

Comment:
We likely want methods to make smaller updates too that are less gas intenstive, like, to update/add/remove a single validator.